docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,391 @@
+```go
+type Operation string
+
+const (
+	WriteOp     Operation = "write"
+	ReadOp      Operation = "read"
+	DeleteOp    Operation = "delete"
+	IterKeyOp   Operation = "iterKey"
+	IterValueOp Operation = "iterValue"
+)
+
+// TraceOperation implements a traced KVStore operation
+type TraceOperation struct {
+	Operation Operation              `json:"operation"`
+	Key       string                 `json:"key"`
+	Value     string                 `json:"value"`
+	Metadata  map[string]interface{} `json:"metadata"`
+}
+```
+
+### Listener type
+We will create a concrete implementation of the `Listening` interface in `store/types/listening.go`.
+This implementation will be configurable with a list of allowed `Operation`s, a set of whitelisted keys and prefixes, and a set of blacklisted keys and prefixes.
+
+```go
+
+// Listener is used to configure listening on specific keys of a KVStore
+type Listener struct {
+	writer              io.Writer
+	context             TraceContext
+	allowedOperations   map[Operation]struct{} // The operations which this listener is allowed to listen to
+	whitelistedKeys     [][]byte               // Keys explicitly allowed to be listened to
+	blacklistedKeys     [][]byte               // Keys explicitly disallowed to be listened to
+	whitelistedPrefixes [][]byte               // Key prefixes explicitly allowed to be listened to
+	blacklistedPrefixes [][]byte               // Key prefixes explicitly disallowed to be listened to
+}
+
+...
+
+// GetContext satisfies Listening interface
+func (l *Listener) GetContext() TraceContext {
+	return l.context
+}
+
+// Write satisfies Listening interface
+// it wraps around the underlying writer interface
+func (l *Listener) Write(b []byte) (int, error) {
+	return l.writer.Write(b)
+}
+
+// Allowed satisfies Listening interface
+// it returns whether or not the Listener is allowed to listen to the provided operation at the provided key
+func (l *Listener) Allowed(op Operation, key []byte) bool {
+	// first check if the operation is allowed
+	if _, ok := l.allowedOperations[op]; !ok {
+		return false
+	}
+	// if there are no keys or prefixes in the whitelists then every key is allowed (unless disallowed in blacklists)
+	// if there are whitelisted keys or prefixes then only the keys which conform are allowed (unless disallowed in blacklists)
+	allowed := true
+	if len(l.whitelistedKeys) > 0 || len(l.whitelistedPrefixes) > 0 {
+		allowed = listsContain(l.whitelistedKeys, l.whitelistedPrefixes, key)
+	}
+	return allowed && !listsContain(l.blacklistedKeys, l.blacklistedPrefixes, key)
+}
+
+func listsContain(keys, prefixes [][]byte, key []byte) bool {
+	for _, k := range keys {
+		if bytes.Equal(key, k) {
+			return true
+		}
+	}
+	for _, p := range prefixes {
+		if bytes.HasPrefix(key, p) {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
This should use maps rather than linear search.

Reply:
Sounds good, the trade off is that we need to convert to strings since we can't use slices as map keys.

Reply:
So let me ask you this, for just the use case of caching to an external database will you be watching all keys or just some keys? i.e. is the use-case for watching some keys more related to gRPC subscribers?

Either way, I'm not really comfortable with watches on resources that are high churn being dealt with using linear search. Keep in mind that the way this is designed currently, it blocks writes on every single store write. This is a _very_ high cost.

I would say for the use case of caching to an external DB we really just want to capture _all_ writes with fault tolerance to some sort of on-disk WAL and do it fast to not block writing too much. If writing is blocked too much, the sync process simply won't be able to keep up when there's heavy usage. The sync process should then consume that WAL on a separate thread as quickly as possible.

For gRPC subscribers which I'm assuming might generally be web clients, I'm not sure fault tolerance is so important. Either way, the dispatch of all that data should also happen in separate threads so that writing is not blocked too much.

For any of these cases, if we're doing filtering for one client or many, I don't really think this is a use case where we settle for an inefficient design. In the web-client case, we should assume many clients, and in all cases we should imagine heavy writes. I actually think the web-client case should be dealt with separately because responding to many clients seems like its own problem space. In any event, I think we'll need to think beyond the restrictions of go `map`s and consider tries, etc. Maybe there are off the shelf solutions that would do that sort of subscription filtering for many clients. Anyway, for now maybe let's try to just get the database sync solution right?

Reply:
Thank you, that helps put things in perspective! Yeah for caching and indexing in a database we will be watching every key, I can think of hypothetical scenarios where we would only want to cache and index a subset of the data but that is both out of scope and, well, hypothetical.

For the stock  `WriteListener` I'll remove the filtering, If someone wants to filter they can implement a `WriteListener` that does so (using a more efficient approach for the prefixes than the one here) or they can perform the filtering at a higher level.

Reply:
Great! I think that simplifies a lot what needs to be done for this first iteration, right?

Reply:
Definitely!